tracing/src/main/java/com/palantir/tracing/DeferredTracer.java

```diff
@@ -151,6 +166,25 @@ public void close() {
         }
     }
 
+    private static final class TaggedCloseableTrace implements CloseableTrace {
+        private final Map<String, String> metadata;
+        private final Optional<Trace> originalTrace;
+
+        TaggedCloseableTrace(Optional<Trace> originalTrace, Map<String, String> metadata) {
+            this.metadata = metadata;
+            this.originalTrace = originalTrace;
+        }
+
+        @Override
+        public void close() {
+            Tracer.fastCompleteSpan(metadata);
```

Comment:
I think this method can be simplified to avoid the `getAndClear` branch when an original trace is present:
```java
Tracer.fastCompleteSpan(metadata);
if (originalTrace.isPresent()) {
    Tracer.setTrace(originalTrace.get());
} else if (Tracer.hasTraceId()) {
    Tracer.getAndClearTrace();
}
```